tokenscan/scanner.go

```diff
@@ -0,0 +1,221 @@
+package tokenscan
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gocql/gocql"
+
+	"github.com/scylladb/gocqlx"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/scylladb/gocqlx/qb"
+)
+
+const (
+	// parallelism = nodesInCluster * coresInNode * 3
+	defaultParallelism = 72
+)
+
+type tokenRange struct {
+	startRange int64
+	endRange   int64
+}
+
+// SessionBuilder creates a gocql.Session
+type SessionBuilder interface {
+	CreateSession() (*gocql.Session, error)
+}
+
+// ApplyFunc is a callback function that is called with each retrieved scan item.
+// The scan is aborted if the ApplyFunc returns false
+type ApplyFunc func(context.Context, map[string]interface{}) bool
+
+// Scanner is a scylla token range scanner
+type Scanner struct {
```

Comment:
I doubt shis struct is need - Having just a function would be easier.

Reply:
I implemented it as a struct to be able to implement the `prometheus.Collector` interface on it. If we don't want this (which I can understand to not pull in too many dependencies), I can surely implement the scanner as a function only.

On the other hand the constructor returns no error, so the call of the constructor + Scan function can be chained, so it doesn't really make a difference.

Reply:
Single function is a better API, however, there can be an unexported scanner type to reduce worker and producer scope and to allow for code separation - unspaghetting :)